src/config/id.rs

```diff
@@ -7,43 +7,96 @@ use std::{
     fmt,
 };
 
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum ComponentScope {
+    Global,
+    Pipeline(String),
+}
+
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
 pub struct ComponentId {
+    value: String,
     id: String,
+    scope: ComponentScope,
 }
 
 impl ComponentId {
     pub fn global<T: Into<String>>(id: T) -> Self {
-        Self { id: id.into() }
+        let id = id.into();
+        Self {
+            id: id.clone(),
+            value: id,
+            scope: ComponentScope::Global,
+        }
     }
 
-    pub fn as_str(&self) -> &str {
+    pub fn pipeline(pipeline: &str, id: &str) -> Self {
+        let value = format!("{}#{}", pipeline, id);
+        Self {
+            id: id.to_string(),
+            value,
+            scope: ComponentScope::Pipeline(pipeline.to_string()),
+        }
+    }
+
+    pub fn id(&self) -> &str {
         self.id.as_str()
     }
+
+    pub fn pipeline_str(&self) -> Option<&str> {
+        match self.scope {
+            ComponentScope::Pipeline(ref value) => Some(value.as_str()),
+            _ => None,
+        }
+    }
+
+    pub fn into_pipeline(self, id: &str) -> Self {
+        Self::pipeline(id, &self.id)
+    }
+
+    pub fn is_global(&self) -> bool {
+        matches!(self.scope, ComponentScope::Global)
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.value.as_str()
+    }
 }
+
 impl From<String> for ComponentId {
     fn from(value: String) -> Self {
-        Self { id: value }
+        Self::from(value.as_str())
     }
 }
 
 impl From<&str> for ComponentId {
     fn from(value: &str) -> Self {
-        Self::from(value.to_string())
+        let parts = value.split('#').take(2).collect::<Vec<_>>();
```

Comment:
Just a note for posterity that we discussed in stand-up making this a `.`. We also discussed validating configs to ensure that users are not using a `.` already in their component id.